pyeudiw/trust/dynamic.py

```diff
@@ -1,100 +1,46 @@
-    trust sources are queried when some metadata or key material is requested.
-    """
-
-    def __init__(self, trust_evaluators: dict[str, TrustEvaluator], storage: Optional[DBEngine] = None):
-        self.trust_evaluators: dict[str, TrustEvaluator] = trust_evaluators
-        self.storage: DBEngine | None = storage
-
-    def _get_trust_identifier_names(self) -> str:
-        return f'[{",".join(self.trust_evaluators.keys())}]'
-
-    def _get_public_keys_from_storage(self, eval_identifier: str, issuer: str) -> list[dict] | None:
-        """
-        Search public key for trust model 'eval_identifier' in the storage layer (if any). If the storage
-        layer fails or does not exists, None is returned.
-        Public keys are intended to be serialized as jwks (jwk set) in the storage layer.
-
-        :returns: a JWKS dictionary if the keys are found in the storage, or None if storage lookup fails
-        :rtype: dict | None
-        """
-        if not self.storage:
-            return None
-
-        if trust_attestation := self.storage.get_trust_attestation(issuer):
-            if trust_entity := trust_attestation.get(eval_identifier, None):
-                if trust_entity_jwks := trust_entity.get("jwks", None):
-                    new_pks = trust_entity_jwks
-                    # TODO: check if cached key is still valid?
-                    # with mongodb we use ttl integrated in the engine
-                    return new_pks
-        return None
-
-    def _get_public_keys(self, eval_identifier: str, eval_instance: TrustEvaluator, issuer: str) -> list[dict]:
-        new_pks: list = []
+    def __init__(self, handlers: list[TrustHandlerInterface], db_engine: DBEngine) -> None:
+        self.db_engine: DBEngine = db_engine
+        self.handlers: list[TrustHandlerInterface] = handlers
+        self.handlers_names: list[str] = [e.name for e in self.handlers]
+    
+    def _retrieve_trust_source(self, issuer: str) -> Optional[TrustSourceData]:
         try:
-            new_pks = eval_instance.get_public_keys(issuer)
-            if self.storage:
-                self.storage.add_or_update_trust_attestation(issuer, trust_type=TrustType(eval_identifier), jwks=new_pks)
-        except Exception:
-            new_pks = self._get_public_keys_from_storage(eval_identifier, issuer)
-
-        if new_pks:
-            return new_pks
-        raise Exception(f"unable to find any public key with trust model {eval_identifier}")
+            trust_source = self.db_engine.get_trust_source(issuer)
+            return TrustSourceData.from_dict(trust_source)
+        except EntryNotFound:
+            return None
+    
+    def _extract_trust_source(self, issuer: str) -> Optional[TrustSourceData]:
+        trust_source = TrustSourceData.empty(issuer)
+
+        for extractor in self.handlers:
+            trust_source = extractor.extract(issuer, trust_source)
+        
+        self.db_engine.add_trust_source(trust_source.serialize())
```

Comment:
Q: should this line be inside the for loop? As is the code only persists in the DB only the last trust source

Reply:
UPDATE: further inspection suggests that the purpose of the `extract` method is to mutate the trust source received as input so that it is expanded with new material found by the TrustHandler.

If this is the case, then my previous comment is not correct.

Reply:
Resolved in [33c0a54](https://github.com/italia/eudi-wallet-it-python/pull/284/commits/33c0a54145e85bc387e4b4420245fc4af56abac8)